Frameworks/UIKit/UIFontDescriptor.mm

```diff
@@ -29,233 +36,176 @@
-/**@Status Stub */
-const CGFloat UIFontWeightBold = StubConstant();
-/**@Status Stub */
-const CGFloat UIFontWeightHeavy = StubConstant();
-/**@Status Stub */
-const CGFloat UIFontWeightBlack = StubConstant();
-
-@implementation UIFontDescriptor {
-@private
-    idretaintype(NSString) _name;
-    float _size;
-    UIFontDescriptorSymbolicTraits _traits;
-}
+const CGFloat UIFontWeightUltraLight = -0.80f;
+const CGFloat UIFontWeightThin = -0.60f;
+const CGFloat UIFontWeightLight = -0.40f;
+const CGFloat UIFontWeightRegular = 0.00f;
+const CGFloat UIFontWeightMedium = 0.23f;
+const CGFloat UIFontWeightSemibold = 0.30f;
+const CGFloat UIFontWeightBold = 0.40f;
+const CGFloat UIFontWeightHeavy = 0.56f;
+const CGFloat UIFontWeightBlack = 0.62f;
 
-/**
- @Status Interoperable
-*/
-+ (UIFontDescriptor*)fontDescriptorWithName:(NSString*)fontName size:(CGFloat)size {
-    UIFontDescriptor* fontDescriptor = [[UIFontDescriptor alloc] init];
-    fontDescriptor->_name = [UIFontDescriptor _getFallBackFontName:fontName];
-    fontDescriptor->_size = [UIFontDescriptor _getFallBackFontSize:size];
-    return [fontDescriptor autorelease];
-}
+@implementation UIFontDescriptor
 
-+ (UIFontDescriptor*)fontDescriptorWithDescriptor:(UIFontDescriptor*)descriptor size:(CGFloat)size {
-    UIFontDescriptor* fontDescriptor = [[UIFontDescriptor alloc] init];
-    fontDescriptor->_name = descriptor->_name;
-    fontDescriptor->_size = [UIFontDescriptor _getFallBackFontSize:size];
-    fontDescriptor->_traits = descriptor->_traits;
-    return [fontDescriptor autorelease];
-}
+BASE_CLASS_REQUIRED_IMPLS(UIFontDescriptor, UIFontDescriptorPrototype, CTFontDescriptorGetTypeID);
 
 /**
  @Status Caveat
-  @Notes         Force to use default font name and size.
+ @Notes  Force to use default font name and size.
 */
 + (UIFontDescriptor*)preferredFontDescriptorWithTextStyle:(NSString*)style {
     return [UIFontDescriptor fontDescriptorWithName:nil size:0.0];
 }
 
 /**
- This is a temp solution, we need to implement the fall back logic in 1512.
- Here just to check nil, and return default system font as fall back.
- */
-+ (NSString*)_getFallBackFontName:(NSString*)fontName {
-    NSString* ret = [NSString alloc];
-    if (fontName == nil || fontName.length == 0) {
-        ret = [ret initWithString:SystemFontName];
-    } else {
-        ret = [ret initWithString:fontName];
-    }
-    return [ret autorelease];
-}
-
-+ (float)_getFallBackFontSize:(float)size {
-    if (size <= 0) {
-        return kCTFontSystemFontSize;
-    } else {
-        return size;
-    }
-}
-
-/**
  @Status Interoperable
 */
-- (UIFontDescriptor*)fontDescriptorWithSymbolicTraits:(UIFontDescriptorSymbolicTraits)symbolicTraits {
-    UIFontDescriptor* result = [UIFontDescriptor fontDescriptorWithDescriptor:self size:self->_size];
-    result->_traits = symbolicTraits;
-    return result;
++ (UIFontDescriptor*)fontDescriptorWithFontAttributes:(NSDictionary*)attributes {
```

Comment:
I ended up reordering the functions in this file to match that of the documentation - the old ordering sort of made sense in the old impl, but is very disorganized in the new one.